### PR TITLE
fix(rpc): use max gas limit if no transaction gas is specified

### LINKED ISSUE #3218: `trace_call` uses incorrect gas limit when it's not provided in the input
### Describe the bug

When no `gas` is provided in the `trace_call` tx object (first input argument), we derive it from the block the call is executed at (third input argument): https://github.com/paradigmxyz/reth/blob/d6092cf1d3bf6dcd274815ce12c15726293a529b/crates/rpc/rpc/src/eth/api/transactions.rs#L541 while both geth and Erigon uses the pre-configured constant of 50 million:
https://github.com/ledgerwatch/erigon/blob/eae2d9a79cb70dbe30b3a6b79c436872e4605458/cmd/rpcdaemon/commands/trace_adhoc.go#L956
https://github.com/ledgerwatch/erigon/blob/eae2d9a79cb70dbe30b3a6b79c436872e4605458/eth/ethconfig/config.go#L94

Reth's behaviour seems to be logically correct, as we inherit the whole environment from the specified block we trace the call at. But it's not consistent with what geth and Erigon currently do.

### Steps to reproduce

1. Start reth and erigon
2. Call `trace_call`:
```json
{
  "jsonrpc": "2.0",
  "method": "trace_call",
  "params": [
    {
      "to": "0x6b175474e89094c44da98b954eedeac495271d0f",
      "data": "0x70a082310000000000000000000000005d3a536e4d6dbd6114cc1ead35777bab948e3643"
    },
    [
      "trace"
    ],
    "0xe67702"
  ],
  "id": 314067518751704834
}
```

3. Observe the difference in responses:
```
type                index       reth       erigon
──────────────────────────────────────────────────
value   trace.0.action.gas   0x1c8fd59   0x2fa9cc8
                              29949273   49978568
```

reth's `29949273` gas limit is the mainnet block 15103746 gas limit `29970705` minus the `21432` gas (min tx gas + initialization costs)
erigon's `49978568` gas limit is the constant gas limit `50000000` minus the same `21432` gas

### Code of Conduct

- [X] I agree to follow the Code of Conduct

### PR DESCRIPTION
**Pull request by shekhirin**
Resolves https://github.com/paradigmxyz/reth/issues/3218

This PR makes all call RPC methods (`eth_call` and call tracing in both `debug` and `trace` namespaces) to set the transaction's (in which the call is executed) gas limit to the constant 30 million if no `gas` is specified in the transaction arguments. Previously, we were deriving it from the block the call is executing at.

Transaction/block tracing methods behaviour is left unchanged.

### CODE REVIEW

File: crates/rpc/rpc/src/eth/api/transactions.rs [hunk 1]
```diff
@@ -263,21 +288,28 @@ where
...
             let block_hash = self
                 .provider()
                 .block_hash_for_id(at)?
                 .ok_or_else(|| EthApiError::UnknownBlockNumber)?;
             let (cfg, env) = self.cache().get_evm_env(block_hash).await?;
-            Ok((cfg, env, block_hash.into()))
+            (cfg, env, block_hash.into())
+        };
+
+        if matches!(block_gas_limit, BlockGasLimit::Max) {
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/api/transactions.rs:302]: I prefer helper functions for this, simple `is_max(`

**Review by mattsse — COMMENTED** [review]
I don't fully understand this PR because this touches block gas limit, not the tx gas limit

and `MAX` and `Header` seem to be the same 30M?

**Comment by shekhirin** [comment]
> this touches block gas limit, not the tx gas limit

We use block and its gas limit to prepare the revm env the call is executed in: https://github.com/paradigmxyz/reth/blob/6b7fe259bc0389d0d36385b07f40f660c2404b0c/crates/rpc/rpc/src/eth/api/transactions.rs#L573-L577

> and MAX and Header seem to be the same 30M?

First, we get the actual block and its gas limit from the database: https://github.com/paradigmxyz/reth/blob/6b7fe259bc0389d0d36385b07f40f660c2404b0c/crates/rpc/rpc/src/eth/api/transactions.rs#L298

Then, if the requested block gas limit should be maximum (30M), we set it to so, overriding the real one: https://github.com/paradigmxyz/reth/blob/6b7fe259bc0389d0d36385b07f40f660c2404b0c/crates/rpc/rpc/src/eth/api/transactions.rs#L302-L304

File: crates/rpc/rpc/src/eth/revm_utils.rs [hunk 2]
```diff
@@ -236,10 +237,21 @@ where
...
+            trace!(target: "rpc::eth::call", ?env, "Applying gas limit cap with caller allowance");
+            cap_tx_gas_limit_with_caller_allowance(db, &mut env.tx)?;
+        } else {
+            // If no gas price is specified, use maximum allowed gas limit. The reason for this is
+            // that both Erigon and Geth use pre-configured gas cap even if it's possible
+            // to derive the gas limit from the block:
+            // https://github.com/ledgerwatch/erigon/blob/eae2d9a79cb70dbe30b3a6b79c436872e4605458/cmd/rpcdaemon/commands/trace_adhoc.go#L956
+            // https://github.com/ledgerwatch/erigon/blob/eae2d9a79cb70dbe30b3a6b79c436872e4605458/eth/ethconfig/config.go#L94
+            trace!(target: "rpc::eth::call", ?env, "Applying gas limit cap as the maximum gas limit");
+            env.tx.gas_limit = ETHEREUM_BLOCK_GAS_LIMIT;
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/revm_utils.rs:253]: this seems okay for now, we could make this configurable, because geth has an option to set this via CLI

**Review by mattsse — APPROVED** [review]

File: crates/rpc/rpc/src/eth/revm_utils.rs (hunk 2 above)
Comment (shekhirin) [reply crates/rpc/rpc/src/eth/revm_utils.rs:253]: Good point, and we should use that configurable gas cap here (in calls) and in gas estimation: https://github.com/ethereum/go-ethereum/blob/5948a3aa2348a53707e8032c99ec28070d870f41/internal/ethapi/api.go#L1158-L1163

Comment (shekhirin) [reply crates/rpc/rpc/src/eth/revm_utils.rs:253]: Will open a good first issue

**Pull request merged**